File: UnitC/ch06/data-6.3-crows-never-forget-face-proportion.jl
#=
Crows Never Forget a Face
locks5 stat page 473

乌鸦能记住人脸吗?
当生物学家为了研究乌鸦的行为, 需要抓住乌鸦然后为其戴上脚环标签,乌鸦似乎非常情愿这么做,
所以会对着生物学家大声鸣叫,似乎过很长时间这些乌鸦也能认出当初为其打标记的生物学家,没有打过标签的乌鸦似乎也对生物学家有反应
因此乌鸦可能还有相互交流人脸的能力.

测试乌鸦是否对戴着野人面具和普通面具叫声的比例

戴着 caveman mask  的人 为乌鸦打标记 
戴着  neutral mask 的人 作为对照

使用   FisherExactTest

=#

## 1. load package
include("../../utils.jl")


## 2. produce data 
df=(let
c2=[158,444-158];c3=[109,922-109]
typename=["scold","no scold","Total"]
df=freq_table(DataFrame("cavemanmask(caught crows)"=>c2,"neutralmask(control)"=>c3);typename=typename)
df end)
 """
┌──────────┬───────────────────────────┬──────────────────────┬───────┐
│     Type │ cavemanmask(caught crows) │ neutralmask(control) │ Total │
│   String │                     Int64 │                Int64 │ Int64 │
├──────────┼───────────────────────────┼──────────────────────┼───────┤
│    scold │                       158 │                  109 │   267 │
│ no scold │                       444 │                  922 │  1366 │
│    Total │                       602 │                 1031 │  1633 │
└──────────┴───────────────────────────┴──────────────────────┴───────┘
"""
a=df[1,2]  #  scold with caveman mask
c=df[3,2] # total  caveman mask

b=df[1,3] # scold with  neutral mask
d=df[3,3] # total  neutral mask
#a/c  scold with caveman mask  rate , scold with neutral mask rate
ht=FisherExactTest(a,b,c,d)

#= 
    Fisher's exact test
-------------------
Population details:
    parameter of interest:   Odds ratio
    value under h_0:         1.0
    point estimate:          3.00788
    95% confidence interval: (2.28, 3.979)

Test summary:
    outcome with 95% confidence: reject h_0
    two-sided p-value:           <1e-15

Details:
    contingency table:
        158  109
        444  922
=#
confint(ht;level=0.90)

File: marimo/ch06/data-6.3-crows-never-forget-face-proportion.py
import marimo

__generated_with = "0.24.0"
app = marimo.App(width="medium")


@app.cell
def _():
    import marimo as mo
    import sys
    from pathlib import Path
    import numpy as np
    import polars as pl
    import matplotlib.pyplot as plt
    from scipy import stats
    from great_tables import GT
    _p = Path(mo.notebook_location()).resolve().parent
    while _p != _p.parent and not (_p / "lock5.py").exists():
        _p = _p.parent
    if str(_p) not in sys.path:
        sys.path.insert(0, str(_p))
    import lock5
    return mo, sys, Path, np, pl, plt, stats, GT, lock5


@app.cell
def _(mo):
    mo.md(r"""# Data 6.3 — Crows Never Forget a Face (Fisher's exact test)

Port of `UnitC/ch06/data-6.3-crows-never-forget-face-proportion.jl`.

The book's counts are 158 of 602 crows scolded the caveman mask and 109 of 1031
scolded the neutral mask (i.e. ``no scold`` = 444 and 922).  The original Julia
source wrote ``444-158`` / ``922-109`` (a bug); this port uses the correct counts.""")
    return


@app.cell
def _(lock5, pl):
    c2 = [158, 444]
    c3 = [109, 922]
    df = lock5.freq_table(
        pl.DataFrame({"cavemanmask(caught crows)": c2, "neutralmask(control)": c3}),
        typename=["scold", "no scold", "Total"],
    )
    return df


@app.cell
def _(GT, df, mo):
    mo.Html(GT(df).tab_header(title="Crow scolding counts").as_raw_html())
    return


@app.cell
def _(df, lock5, mo):
    a = df[0, 1]
    b = df[0, 2]
    c = df[2, 1]
    d = df[2, 2]
    ht = lock5.fisher_exact_test(a, b, c, d)
    mo.Html(lock5.gt_ttest(ht).as_raw_html())
    return a, b, c, d, ht


@app.cell
def _(GT, a, b, c, d, lock5, mo, pl):
    ht90 = lock5.fisher_exact_test(a, b, c, d, level=0.90)
    ci90_df = pl.DataFrame({
        "90% confidence interval": [f"({round(ht90.ci_low, 4)}, {round(ht90.ci_high, 4)})"],
    })
    mo.Html(GT(ci90_df).tab_header(title="Odds ratio — 90% CI").as_raw_html())
    return ht90


if __name__ == "__main__":
    app.run()
